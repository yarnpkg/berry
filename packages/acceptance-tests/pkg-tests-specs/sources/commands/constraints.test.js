const {
  fs: {writeFile},
} = require(`pkg-tests-core`);

const {environments} = require(`./constraints/environments`);

const constraints = {
  [`empty constraints`]: ``,
  [`gen_enforced_dependency (missing)`]: `gen_enforced_dependency(WorkspaceCwd, 'one-fixed-dep', '1.0.0', peerDependencies).`,
  [`gen_enforced_dependency (incompatible)`]: `gen_enforced_dependency(WorkspaceCwd, 'no-deps', '2.0.0', dependencies).`,
  [`gen_enforced_dependency (extraneous)`]: `gen_enforced_dependency(WorkspaceCwd, 'no-deps', null, _).`,
  [`gen_enforced_dependency (extraneous2)`]:
    `
    gen_enforced_dependency(WorkspaceCwd, 'no-deps', null, _) :-
      WorkspaceCwd \\= '.'.
    gen_enforced_dependency(WorkspaceCwd, 'no-deps', '1.0.0', DependencyType) :- 
      workspace_has_dependency(WorkspaceCwd, 'no-deps', '1.0.0', DependencyType).
    `,
  [`gen_enforced_dependency (ambiguous)`]: `gen_enforced_dependency(WorkspaceCwd, 'no-deps', '1.0.0', dependencies). gen_enforced_dependency(WorkspaceCwd, 'no-deps', '2.0.0', dependencies).`,
  [`gen_enforced_field (missing)`]: `gen_enforced_field(WorkspaceCwd, 'dependencies["a-new-dep"]', '1.0.0').`,
  [`gen_enforced_field (incompatible)`]: `gen_enforced_field(WorkspaceCwd, 'dependencies["no-deps"]', '2.0.0').`,
  [`gen_enforced_field (extraneous)`]: `gen_enforced_field(WorkspaceCwd, 'dependencies', null).`,
  [`gen_enforced_field (ambiguous)`]: `gen_enforced_field(WorkspaceCwd, 'dependencies["a-new-dep"]', '1.0.0'). gen_enforced_field(WorkspaceCwd, 'dependencies["a-new-dep"]', '2.0.0').`,
};

describe(`Commands`, () => {
  describe(`constraints`, () => {
    for (const [environmentDescription, environment] of Object.entries(environments)) {
      for (const [scriptDescription, script] of Object.entries(constraints)) {
        test(`test (${environmentDescription} / ${scriptDescription})`,
          makeTemporaryEnv({}, {
            plugins: [
              require.resolve(`@yarnpkg/monorepo/scripts/plugin-constraints.js`),
            ],
          }, async ({path, run, source}) => {
            await environment(path);
            await writeFile(`${path}/constraints.pro`, script);

            let code;
            let stdout;
            let stderr;

            try {
              ({code, stdout, stderr} = await run(`constraints`));
            } catch (error) {
              ({code, stdout, stderr} = error);
            }

            expect({code, stdout, stderr}).toMatchSnapshot();
          }),
        );
      }
    }
  });
});
