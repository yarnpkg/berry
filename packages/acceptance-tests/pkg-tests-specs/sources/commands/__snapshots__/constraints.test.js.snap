// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands constraints test (empty project / empty constraints) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (empty project / gen_enforced_dependency (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (empty project / gen_enforced_dependency (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (empty project / gen_enforced_dependency (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (empty project / gen_enforced_dependency (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (empty project / gen_enforced_field (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (empty project / gen_enforced_field (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (empty project / gen_enforced_field (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (empty project / gen_enforced_field (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / empty constraints) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / gen_enforced_dependency (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / gen_enforced_dependency (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / gen_enforced_dependency (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / gen_enforced_dependency (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / gen_enforced_field (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / gen_enforced_field (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / gen_enforced_field (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (multiple workspaces / gen_enforced_field (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / empty constraints) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / gen_enforced_dependency (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / gen_enforced_dependency (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / gen_enforced_dependency (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / gen_enforced_dependency (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / gen_enforced_field (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / gen_enforced_field (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / gen_enforced_field (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (one regular dependency / gen_enforced_field (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / empty constraints) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / gen_enforced_dependency (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / gen_enforced_dependency (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / gen_enforced_dependency (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / gen_enforced_dependency (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / gen_enforced_field (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / gen_enforced_field (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / gen_enforced_field (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two development dependencies / gen_enforced_field (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / empty constraints) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / gen_enforced_dependency (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / gen_enforced_dependency (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / gen_enforced_dependency (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / gen_enforced_dependency (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / gen_enforced_field (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / gen_enforced_field (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / gen_enforced_field (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies / gen_enforced_field (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / empty constraints) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / gen_enforced_dependency (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / gen_enforced_dependency (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / gen_enforced_dependency (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / gen_enforced_dependency (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / gen_enforced_field (ambiguous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / gen_enforced_field (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / gen_enforced_field (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;

exports[`Commands constraints test (two regular dependencies, two development dependencies / gen_enforced_field (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Unknown Syntax Error: Extraneous positional argument (\\"check\\").

$ yarn constraints [--fix]
",
}
`;
